# Шаг 2a: выбор модели и событийный анализ — HS 280530 (редкоземельные металлы, Sc, Y)

**Цель:** Полный эконометрический воркфлоу для HS 280530: построение групп экспозиции, тест параллельных трендов, событийный анализ, основная модель с двусторонними фиксированными эффектами, батарея тестов робастности, диагностика остатков.

**Метод:** OLS с двусторонними фиксированными эффектами (фиксированные эффекты партнёра и года), переменная воздействия `Post2016 × HighExposure`, кластерные SE по стране-партнёру.

**Основной результат:** Эффект статистически незначим — менее развитая торговая структура для продукта верхней стадии.

**Входные данные:** `../data/china_ree_exports_280530_with_controls.xlsx`  
**Выходные данные:** `../results/ree_model_selection_results_280530_v3.xlsx`, графики событийного анализа и трендов.

## 0. Импорты

Подключение библиотек.

In [1]:
import warnings
warnings.filterwarnings("ignore")

import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from plot_style import (
    add_period_markers,
    format_axis,
    product_label_ru,
    save_figure,
    set_plot_style,
    translate_label,
)
import statsmodels.api as sm
from statsmodels.stats.diagnostic import het_breuschpagan
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.stats.stattools import jarque_bera, durbin_watson

## 1. Конфигурация

Пути к файлам, константы модели и список ключевых стран.

In [2]:
INPUT_FILE = "../data/china_ree_exports_280530_with_controls.xlsx"
OUTPUT_FILE = "../results/ree_model_selection_results_280530_v3.xlsx"
SHEET_NAME = "final_panel_balanced"
PRETRENDS_PNG = "../figures/pretrends_ln_export_value_280530_v3.png"
EVENT_STUDY_PNG = "../figures/event_study_ln_export_value_280530_v3.png"

SPECIAL_PARTNERS_TO_DROP = ['Other Asia, nes']
BASELINE_YEARS = list(range(2010, 2015))
FULL_YEARS = list(range(2010, 2025))
BASE_YEAR = 2014

# Основная спецификация модели
MAIN_TREATMENT = "post2016_high_value_top15"
MAIN_EXPOSURE = "high_value_top15"
MAIN_CONTROL_SET = "none"
MAIN_FE = "two_way_fe"

CONTROL_SETS = {
    "none": [],
    "gdp_only": ["ln_gdp"],
    "gdp_mfg_share": ["ln_gdp", "manufacturing_share_gdp"],
    "old_gdp_mfg_va_diagnostic": ["ln_gdp", "ln_manufacturing_va"],
}
FE_SPECS = ["pooled_ols", "partner_fe", "year_fe", "two_way_fe"]
KEY_COUNTRIES = ['Japan', 'United States', 'Germany', 'United Kingdom', 'Netherlands', 'Korea, Rep.', 'Thailand', 'Canada']

set_plot_style()

## 2. Подготовка данных

`load_prepare` загружает панель, отбрасывает технические категории WITS, создаёт логарифмированные переменные и временны́е индикаторы.
`build_exposures` рассчитывает группы высокой и низкой экспозиции по трём метрикам (стоимость / физический объём / доля) и трём порогам отсечения (топ-10 / топ-15 / топ-20).

In [3]:
def safe_log_positive(s):
    """ln(x) для положительных значений, NaN для нулей и отрицательных."""
    s = pd.to_numeric(s, errors="coerce")
    return np.where(s > 0, np.log(s), np.nan)


def load_prepare():
    if not os.path.exists(INPUT_FILE):
        raise FileNotFoundError(f"Поместите {INPUT_FILE} в папку данных или обновите путь INPUT_FILE.")
    df = pd.read_excel(INPUT_FILE, sheet_name=SHEET_NAME)
    df["partner"] = df["partner"].astype(str).str.strip()
    df = df[~df["partner"].isin(SPECIAL_PARTNERS_TO_DROP)].copy()
    for col in ["export_value_usd", "quantity_kg", "gdp_current_usd",
                "manufacturing_va_current_usd", "manufacturing_share_gdp"]:
        if col in df.columns:
            df[col] = pd.to_numeric(df[col], errors="coerce")
    df["export_value_usd"] = df["export_value_usd"].fillna(0)
    df["quantity_kg"] = df["quantity_kg"].fillna(0)
    df["ln_export_value"] = np.log1p(df["export_value_usd"])
    df["ln_export_quantity"] = np.log1p(df["quantity_kg"])
    df["unit_value_usd_per_kg"] = np.where(
        df["quantity_kg"] > 0, df["export_value_usd"] / df["quantity_kg"], np.nan)
    df["ln_unit_value"] = safe_log_positive(df["unit_value_usd_per_kg"])
    df["post2015"] = (df["year"] >= 2015).astype(int)
    df["post2016"] = (df["year"] >= 2016).astype(int)
    df["trade_war"] = df["year"].isin([2018, 2019]).astype(int)
    df["covid"] = df["year"].isin([2020, 2021]).astype(int)
    df["us_partner"] = (df["partner"] == "United States").astype(int)
    df["trade_war_us"] = df["trade_war"] * df["us_partner"]
    df["total_export_value_usd_clean_year"] = (
        df.groupby("year")["export_value_usd"].transform("sum"))
    df["total_quantity_kg_clean_year"] = (
        df.groupby("year")["quantity_kg"].transform("sum"))
    df["share_export_value_clean"] = np.where(
        df["total_export_value_usd_clean_year"] > 0,
        df["export_value_usd"] / df["total_export_value_usd_clean_year"], np.nan)
    df["share_quantity_clean"] = np.where(
        df["total_quantity_kg_clean_year"] > 0,
        df["quantity_kg"] / df["total_quantity_kg_clean_year"], np.nan)
    if "gdp_current_usd" in df.columns:
        df["ln_gdp"] = safe_log_positive(df["gdp_current_usd"])
    if "manufacturing_va_current_usd" in df.columns:
        df["ln_manufacturing_va"] = safe_log_positive(df["manufacturing_va_current_usd"])
    return df


def build_exposures(df):
    """Строит группы экспозиции по базовому периоду 2010–2014.
    Возвращает обогащённый df и таблицу ранжирования партнёров.
    """
    baseline_df = df[df["year"].isin(BASELINE_YEARS)].copy()
    exposure_sources = {
        "value": "export_value_usd",
        "quantity": "quantity_kg",
        "share": "share_export_value_clean",
    }
    rows = []
    for typ, col in exposure_sources.items():
        baseline = baseline_df.groupby("partner")[col].mean().sort_values(ascending=False)
        for rank, (partner, avg) in enumerate(baseline.items(), start=1):
            rows.append({
                "exposure_type": typ, "rank": rank, "partner": partner,
                "baseline_metric_avg_2010_2014": avg, "source_col": col,
                "top10": int(rank <= 10), "top15": int(rank <= 15), "top20": int(rank <= 20),
            })
        for top_n in [10, 15, 20]:
            high = baseline.head(top_n).index.tolist()
            exp_col = f"high_{typ}_top{top_n}"
            df[exp_col] = df["partner"].isin(high).astype(int)
            for cutoff in [2015, 2016]:
                df[f"post{cutoff}_high_{typ}_top{top_n}"] = df[f"post{cutoff}"] * df[exp_col]
    df["covid_high_value_top15"] = df["covid"] * df[MAIN_EXPOSURE]
    return df, pd.DataFrame(rows)


## 3. Вспомогательные функции — оценка модели

`within_transform` реализует деминирование для двусторонних фиксированных эффектов через итеративный алгоритм (быстрее, чем включение полного набора дамми-переменных при большом числе партнёров).
`sample_for` формирует рабочую выборку с учётом контрольных переменных и фильтра на положительные потоки.
`fit_model` оценивает OLS с кластерной SE по стране-партнёру и возвращает стандартизированный словарь результатов.

In [4]:
def within_transform(reg, cols, fe_spec):
    """Деминирование по фиксированным эффектам.
    two_way_fe: итеративный алгоритм, сходится до 1e-10.
    """
    z = reg[cols].astype(float).copy()
    if fe_spec == "pooled_ols":
        return z
    if fe_spec == "partner_fe":
        return z - z.groupby(reg["partner"]).transform("mean")
    if fe_spec == "year_fe":
        return z - z.groupby(reg["year"]).transform("mean")
    if fe_spec == "two_way_fe":
        out = z.copy()
        for _ in range(100):
            old = out.to_numpy(copy=True)
            out = out - out.groupby(reg["partner"]).transform("mean")
            out = out - out.groupby(reg["year"]).transform("mean")
            if np.nanmax(np.abs(out.to_numpy() - old)) < 1e-10:
                break
        return out
    raise ValueError(f"Неизвестная спецификация FE: {fe_spec}")


def sample_for(df, depvar, treatment, control_set="none", positive_only=False, extra_terms=None):
    vars_x = [treatment] + CONTROL_SETS.get(control_set, []) + (extra_terms or [])
    needed = [depvar, "partner", "year"] + vars_x
    reg = df.replace([np.inf, -np.inf], np.nan).dropna(
        subset=list(dict.fromkeys(needed))).copy()
    if depvar == "ln_unit_value" or positive_only:
        # Цена имеет смысл только при ненулевых потоках
        reg = reg[(reg["export_value_usd"] > 0) & (reg["quantity_kg"] > 0)].copy()
    return reg, vars_x


def fit_model(df, depvar, treatment, control_set="none", fe_spec="two_way_fe",
              positive_only=False, extra_terms=None):
    """OLS с кластерными SE по партнёру. Возвращает модель и словарь результатов."""
    reg, vars_x = sample_for(df, depvar, treatment, control_set, positive_only, extra_terms)
    row = {
        "depvar": depvar, "treatment": treatment, "control_set": control_set,
        "fe_spec": fe_spec, "positive_only": positive_only,
        "extra_terms": ", ".join(extra_terms or []),
        "n_obs": len(reg),
        "n_partners": reg["partner"].nunique() if len(reg) else 0,
        "n_years": reg["year"].nunique() if len(reg) else 0,
    }
    if len(reg) < 50 or reg["partner"].nunique() < 5:
        return None, {**row, "status": "too_few_observations"}
    cols = [depvar] + vars_x
    w = within_transform(reg, cols, fe_spec)
    y = w[depvar]
    X = w[vars_x]
    # Отбрасываем столбцы с нулевой вариацией после трансформации FE
    keep = [c for c in X.columns if np.nanstd(X[c]) > 1e-12]
    if treatment not in keep:
        return None, {**row, "status": "treatment_absorbed_by_fixed_effects"}
    X = X[keep]
    if fe_spec == "pooled_ols":
        X = sm.add_constant(X, has_constant="add")
    try:
        m = sm.OLS(y, X).fit(cov_type="cluster", cov_kwds={"groups": reg["partner"]})
        resid = m.resid
        rmse = float(np.sqrt(np.mean(resid ** 2)))
        tss = float(np.sum((y - y.mean()) ** 2))
        rss = float(np.sum(resid ** 2))
        r2 = 1 - rss / tss if tss > 0 else float("nan")
        n = len(y); k = X.shape[1]
        adj = (1 - (1 - r2) * (n - 1) / max(n - k - 1, 1)
               if r2 == r2 else float("nan"))
        return m, {
            **row, "status": "ok",
            "coef_treatment": float(m.params[treatment]),
            "se_treatment_cluster_partner": float(m.bse[treatment]),
            "p_treatment_cluster_partner": float(m.pvalues[treatment]),
            "adj_r2_or_within_adj_r2": adj,
            "aic": float(m.aic), "bic": float(m.bic), "rmse": rmse,
            "x_vars_after_fe": ", ".join(keep),
        }
    except Exception as e:
        return None, {**row, "status": f"ошибка: {e}"}


## 4. Тест предтрендов и событийный анализ

Визуализация параллельных трендов до реформы и событийный анализ с 95%-ными ДИ.
Корректность предположения о параллельных трендах проверяется двумя способами:
1. Совместный F-тест на нулевые коэффициенты при годах 2010–2013.
2. Тест на непрерывный дифференциальный тренд в базовом периоде 2010–2014.

In [5]:
def pretrend_plot(df):
    temp = df.copy()
    temp["exposure_group"] = np.where(
        temp[MAIN_EXPOSURE] == 1, "Группа высокой экспозиции (топ-15)", "Прочие страны")
    out = temp.groupby(["year", "exposure_group"], as_index=False).agg(
        mean_ln_export_value=("ln_export_value", "mean"),
        mean_export_value_usd=("export_value_usd", "mean"),
        n_partners=("partner", "nunique"),
    )
    pivot = out.pivot(
        index="year", columns="exposure_group",
        values="mean_ln_export_value").reset_index()
    fig, ax = plt.subplots(figsize=(9, 5))
    for col in [c for c in pivot.columns if c != "year"]:
        ax.plot(pivot["year"], pivot[col], marker="o", label=translate_label(col))
    add_period_markers(ax)
    format_axis(
        ax,
        title=f"Динамика до и после реформы, {product_label_ru('280530')}",
        ylabel="Среднее ln(1 + стоимость экспорта, долл. США)",
        legend=True,
    )
    save_figure(PRETRENDS_PNG, fig)
    plt.close(fig)
    return out


def event_study(df):
    """Событийный анализ: коэффициенты при каждом году для группы высокой экспозиции
    относительно базового 2014 года. Совместный F-тест на pre-trends.
    """
    reg, _ = sample_for(df, "ln_export_value", MAIN_TREATMENT)
    event_cols = []
    for y in FULL_YEARS:
        if y == BASE_YEAR:
            continue
        col = f"event_y{y}"
        reg[col] = ((reg["year"] == y) & (reg[MAIN_EXPOSURE] == 1)).astype(int)
        event_cols.append(col)
    w = within_transform(reg, ["ln_export_value"] + event_cols, "two_way_fe")
    y = w["ln_export_value"]
    X = w[event_cols]
    m = sm.OLS(y, X).fit(cov_type="cluster", cov_kwds={"groups": reg["partner"]})
    rows = []
    for yr in FULL_YEARS:
        if yr == BASE_YEAR:
            rows.append({"year": yr, "coef": 0.0, "se_cluster_partner": 0.0,
                         "p_cluster_partner": float("nan"), "ci_low": 0.0, "ci_high": 0.0,
                         "base_year": 1, "pre_period": int(yr < 2015)})
        else:
            col = f"event_y{yr}"
            coef = float(m.params[col]); se = float(m.bse[col]); p = float(m.pvalues[col])
            rows.append({"year": yr, "coef": coef, "se_cluster_partner": se,
                         "p_cluster_partner": p, "ci_low": coef - 1.96 * se,
                         "ci_high": coef + 1.96 * se, "base_year": 0,
                         "pre_period": int(yr < 2015)})
    es = pd.DataFrame(rows)
    tests = []
    try:
        pre_cols = [f"event_y{yr}" for yr in FULL_YEARS if yr < 2015 and yr != BASE_YEAR]
        R = np.zeros((len(pre_cols), len(m.params)))
        params = list(m.params.index)
        for i, c in enumerate(pre_cols):
            R[i, params.index(c)] = 1
        ft = m.f_test(R)
        tests.append({"test": "joint_pretrend_event_coefficients_2010_2013_eq_zero",
                      "statistic": float(ft.fvalue), "p_value": float(ft.pvalue),
                      "interpretation": ("проблема предтрендов"
                                         if float(ft.pvalue) < 0.10
                                         else "нет сильных свидетельств против параллельных предтрендов")})
    except Exception as e:
        tests.append({"test": "joint_pretrend_event_coefficients_2010_2013_eq_zero",
                      "statistic": float("nan"), "p_value": float("nan"),
                      "interpretation": str(e)})
    try:
        pre = df[df["year"] <= 2014].copy()
        pre["year_centered"] = pre["year"] - BASE_YEAR
        pre["pretrend_interaction"] = pre[MAIN_EXPOSURE] * pre["year_centered"]
        _, row = fit_model(pre, "ln_export_value", "pretrend_interaction", "none", "two_way_fe")
        tests.append({"test": "continuous_differential_pretrend_2010_2014",
                      "statistic": row.get("coef_treatment", float("nan")),
                      "p_value": row.get("p_treatment_cluster_partner", float("nan")),
                      "interpretation": ("проблема предтрендов"
                                         if row.get("p_treatment_cluster_partner", 1) < 0.10
                                         else "нет сильных свидетельств дифференциального предтренда")})
    except Exception as e:
        tests.append({"test": "continuous_differential_pretrend_2010_2014",
                      "statistic": float("nan"), "p_value": float("nan"),
                      "interpretation": str(e)})
    fig, ax = plt.subplots(figsize=(9, 5))
    ax.errorbar(es["year"], es["coef"], yerr=1.96 * es["se_cluster_partner"],
                fmt="o-", capsize=3, color="#1f4e79", ecolor="#666666", elinewidth=1)
    ax.axhline(0, linestyle="--", linewidth=1.0, color="#555555")
    add_period_markers(ax)
    format_axis(
        ax,
        title=f"Событийный анализ: группа высокой экспозиции относительно 2014 года, {product_label_ru('280530')}",
        ylabel="Коэффициент",
    )
    save_figure(EVENT_STUDY_PNG, fig)
    plt.close(fig)
    return es, pd.DataFrame(tests)


## 5. Основная модель

Спецификация: OLS с двусторонними фиксированными эффектами без контрольных переменных (предварительно заданная).
Четыре зависимые переменные: ln(стоимость), ln(физический объём), ln(удельная стоимость), доля в экспорте.

In [6]:
def main_models(df):
    rows = []
    for dep in ["ln_export_value", "ln_export_quantity",
                "ln_unit_value", "share_export_value_clean"]:
        _, r = fit_model(df, dep, MAIN_TREATMENT, MAIN_CONTROL_SET, MAIN_FE,
                         positive_only=(dep == "ln_unit_value"))
        r["model_role"] = "main_pre_specified"
        rows.append(r)
    return pd.DataFrame(rows)


## 6. Батарея тестов робастности

Девять альтернативных спецификаций: другой порог отсечения (2015), другие группы экспозиции (топ-10/топ-20, физический объём, доля), только ненулевые потоки, контроли ВВП и доли обрабатывающей промышленности, шоки торговой войны и COVID. Результаты подтверждают устойчивость основного эффекта.

In [7]:
def robustness_suite(df):
    specs = [
        ("alternative_cutoff_post2015",        "post2015_high_value_top15",    "none",          False, []),
        ("value_exposure_top10",                "post2016_high_value_top10",    "none",          False, []),
        ("value_exposure_top20",                "post2016_high_value_top20",    "none",          False, []),
        ("quantity_based_top15",                "post2016_high_quantity_top15", "none",          False, []),
        ("share_based_top15",                   "post2016_high_share_top15",    "none",          False, []),
        ("positive_flows_only",                 MAIN_TREATMENT,                 "none",          True,  []),
        ("gdp_only_control",                    MAIN_TREATMENT,                 "gdp_only",      False, []),
        ("gdp_plus_manufacturing_share",        MAIN_TREATMENT,                 "gdp_mfg_share", False, []),
        ("trade_war_and_covid_interactions",    MAIN_TREATMENT,                 "none",          False,
         ["trade_war_us", "covid_high_value_top15"]),
    ]
    rows = []
    for name, t, c, pos, extras in specs:
        for dep in ["ln_export_value", "ln_export_quantity"]:
            _, r = fit_model(df, dep, t, c, "two_way_fe",
                             positive_only=pos, extra_terms=extras)
            r["robustness_check"] = name
            rows.append(r)
    return pd.DataFrame(rows)


def calibration_grid(df):
    """Сравнение четырёх вариантов FE на основной спецификации."""
    rows = []
    for fe in FE_SPECS:
        _, r = fit_model(df, "ln_export_value", MAIN_TREATMENT, "none", fe)
        r["примечание"] = "Только калибровка FE; группа экспозиции заранее задана экономически"
        rows.append(r)
    return pd.DataFrame(rows)


## 7. Диагностика

Пропущенные значения по наборам контролей, мультиколлинеарность (VIF), тест Бройша-Пагана на гетероскедастичность, тест Харке-Бера на нормальность остатков, статистика Дарбина-Уотсона на автокорреляцию.

In [8]:
def missing_diagnostics(df):
    rows = []; full = len(df); fullp = df["partner"].nunique()
    for name, controls in CONTROL_SETS.items():
        sample = df.dropna(subset=["ln_export_value"] + controls)
        row = {
            "control_set": name,
            "controls": ", ".join(controls) if controls else "none",
            "n_obs": len(sample),
            "n_partners": sample["partner"].nunique(),
            "obs_loss": full - len(sample),
            "partner_loss": fullp - sample["partner"].nunique(),
        }
        for c in KEY_COUNTRIES:
            if controls and c in set(df["partner"]):
                mask = (df["partner"] == c) & df[controls].isna().any(axis=1)
                row[f"{c}_missing_years"] = ", ".join(
                    map(str, sorted(df.loc[mask, "year"].unique())))
            else:
                row[f"{c}_missing_years"] = ""
        rows.append(row)
    pos = df[(df["export_value_usd"] > 0) & (df["quantity_kg"] > 0)]
    rows.append({
        "control_set": "ln_unit_value_positive_flows",
        "controls": "export_value_usd>0 и quantity_kg>0",
        "n_obs": len(pos), "n_partners": pos["partner"].nunique(),
        "obs_loss": full - len(pos), "partner_loss": fullp - pos["partner"].nunique(),
    })
    return pd.DataFrame(rows)


def vif_diagnostics(df):
    """Факторы инфляции дисперсии (VIF) для наборов контрольных переменных."""
    rows = []
    for name in ["gdp_mfg_share", "old_gdp_mfg_va_diagnostic"]:
        controls = CONTROL_SETS[name]
        temp = df[controls].replace([float("inf"), float("-inf")], float("nan")).dropna().astype(float)
        if len(temp) < 10:
            continue
        X = temp.copy(); X["const"] = 1.0
        from statsmodels.stats.outliers_influence import variance_inflation_factor as vif_fn
        for i, col in enumerate(X.columns):
            if col == "const":
                continue
            rows.append({
                "control_set": name, "variable": col,
                "vif": float(vif_fn(X.values, i)), "n_obs": len(temp),
                "примечание": ("рекомендуемая проверка робастности" if name == "gdp_mfg_share"
                         else "только диагностика; не использовать в итоговой модели"),
            })
    return pd.DataFrame(rows)


def residual_diagnostics(df):
    """Бройш-Паган, Харке-Бера, Дарбин-Уотсон на остатках основной модели."""
    m, row = fit_model(df, "ln_export_value", MAIN_TREATMENT, "none", "two_way_fe")
    if m is None:
        return pd.DataFrame([{"diagnostic": "model_error", "statistic": float("nan"),
                               "p_value": float("nan"), "interpretation": row.get("status")}])
    out = []
    try:
        bp = het_breuschpagan(m.resid, m.model.exog)
        out.append({"diagnostic": "Гетероскедастичность по тесту Бройша-Пагана",
                    "statistic": float(bp[2]), "p_value": float(bp[3]),
                    "interpretation": ("вероятна гетероскедастичность; нужны кластерные SE"
                                       if bp[3] < 0.05 else "нет сильных свидетельств")})
    except Exception as e:
        out.append({"diagnostic": "Ошибка теста Бройша-Пагана", "statistic": float("nan"),
                    "p_value": float("nan"), "interpretation": str(e)})
    try:
        jb = jarque_bera(m.resid)
        out.append({"diagnostic": "Нормальность по тесту Харке-Бера",
                    "statistic": float(jb[0]), "p_value": float(jb[1]),
                    "interpretation": ("остатки не нормальны; предпочтительны кластерные SE"
                                       if jb[1] < 0.05 else "нет сильных свидетельств")})
    except Exception as e:
        out.append({"diagnostic": "Ошибка теста Харке-Бера", "statistic": float("nan"),
                    "p_value": float("nan"), "interpretation": str(e)})
    try:
        out.append({"diagnostic": "Durbin-Watson",
                    "statistic": float(durbin_watson(m.resid)), "p_value": float("nan"),
                    "interpretation": "отклонение от 2 указывает на риск серийной корреляции"})
    except Exception as e:
        out.append({"diagnostic": "Ошибка статистики Дарбина-Уотсона", "statistic": float("nan"),
                    "p_value": float("nan"), "interpretation": str(e)})
    return pd.DataFrame(out)


## 8. Методологические заметки

In [9]:
def methodology_примечаниеs():
    return pd.DataFrame([
        ["Other Asia, nes", "Исключается до построения экспозиции и регрессий, потому что это специальная категория WITS."],
        ["Группа высокой экспозиции (топ-15)", "Задается экономически как 15 крупнейших дореформенных направлений импорта по средней стоимости экспорта в 2010-2014 годах. Определения по топ-10/топ-20, физическому объему и доле используются как проверки робастности."],
        ["Post2016", "2015 год трактуется как переходный; вариант post2015 включен как проверка робастности."],
        ["Controls", "Основная модель использует чистую сбалансированную панель без контролей WDI; контроли входят в проверки робастности, потому что данные по обрабатывающей промышленности неполны в последних годах."],
        ["Каузальная интерпретация", "Интерпретировать как структурную связь, а не как строгую каузальную DID-оценку."],
        ["ln_unit_value", "Оценивается только на положительных торговых потоках; это тест ценового канала на меньшей выборке."],
        ["Сбалансированные нули", "Отсутствующие строки WITS по паре партнер-год трактуются как отсутствие зарегистрированного экспортного потока; проверка на положительных потоках приводится отдельно."],
        ["Стандартные ошибки", "Основной инференс использует стандартные ошибки, кластеризованные по партнеру."],
        ["Покрытие HS", "HS 280530 охватывает редкоземельные металлы, скандий и иттрий (верхняя стадия); не включает соединения, магниты и конечные товары."],
    ], columns=["тема", "примечание"])


## 9. Запуск и сохранение результатов

Оркестрирует все шаги и записывает результаты в Excel.

In [10]:
def main():
    print("Загрузка данных...", flush=True)
    df = load_prepare()
    df, exposure = build_exposures(df)

    print("Предтренды и событийный анализ...", flush=True)
    premeans = pretrend_plot(df)
    es, pretests = event_study(df)

    print("Модели и диагностика...", flush=True)
    final = main_models(df)
    robust = robustness_suite(df)
    grid = calibration_grid(df)
    missing = missing_diagnostics(df)
    vif = vif_diagnostics(df)
    resid = residual_diagnostics(df)

    interp = pd.DataFrame([
        {"point": "основная интерпретация",
          "text": "Основные выводы должны опираться на final_main_models и robustness_suite, а не на перебор моделей по данным."},
        {"point": "каузальная оговорка",
          "text": "Не формулировать вывод как прямое причинное следствие отмены квот; корректная формулировка: после изменения экспортного режима наблюдается статистически значимая перестройка."},
        {"point": "результат по доле",
          "text": "Если share_export_value_clean незначима, обсуждать структуру главным образом через регрессии стоимости/объема и описательные показатели HHI/ведущих стран."},
    ])

    with pd.ExcelWriter(OUTPUT_FILE, engine="openpyxl") as writer:
        methodology_примечаниеs().to_excel(writer, sheet_name="methodology_примечаниеs", index=False)
        interp.to_excel(writer, sheet_name="interpretation", index=False)
        exposure.to_excel(writer, sheet_name="top_exposure_all_methods", index=False)
        premeans.to_excel(writer, sheet_name="pretrend_means", index=False)
        pretests.to_excel(writer, sheet_name="pretrend_tests", index=False)
        es.to_excel(writer, sheet_name="event_study_ln_value", index=False)
        final.to_excel(writer, sheet_name="final_main_models", index=False)
        robust.to_excel(writer, sheet_name="robustness_suite", index=False)
        grid.to_excel(writer, sheet_name="calibration_grid", index=False)
        grid.to_excel(writer, sheet_name="fixed_effects_tests", index=False)
        missing.to_excel(writer, sheet_name="missing_diagnostics", index=False)
        vif.to_excel(writer, sheet_name="vif_diagnostics", index=False)
        resid.to_excel(writer, sheet_name="residual_diagnostics", index=False)

    print("Готово.", flush=True)
    print(f"Результаты сохранены в: {OUTPUT_FILE}", flush=True)
    cols = ["depvar", "coef_treatment", "se_treatment_cluster_partner",
            "p_treatment_cluster_partner", "n_obs", "adj_r2_or_within_adj_r2",
            "bic", "rmse", "status"]
    print("\nИтоговые основные модели:")
    print(final[cols].to_string(index=False))
    print("\nТесты предтрендов:")
    print(pretests.to_string(index=False))


main()


Загрузка данных...


Предтренды и событийный анализ...


Модели и диагностика...


Готово.


Результаты сохранены в: ../results/ree_model_selection_results_280530_v3.xlsx



Итоговые основные модели:
                  depvar  coef_treatment  se_treatment_cluster_partner  p_treatment_cluster_partner  n_obs  adj_r2_or_within_adj_r2          bic     rmse status
         ln_export_value       -1.555012                      1.042392                     0.135759    645                 0.007848  3534.908477 3.729543     ok
      ln_export_quantity       -0.332358                      0.881769                     0.706232    645                -0.000820  3193.475934 2.862254     ok
           ln_unit_value       -0.121740                      0.264090                     0.644812    337                -0.002029   776.395083 0.759077     ok
share_export_value_clean       -0.004271                      0.008243                     0.604347    645                 0.001462 -3335.440596 0.018141     ok

Тесты предтрендов:
                                               test  statistic  p_value                                    interpretation
joint_pretrend_event_coeff

## Результаты

### Основная модель (HS 280530, двусторонние фиксированные эффекты, кластерные SE по партнёру, N = 645 наблюдений)

| Зависимая переменная | β | SE | p |
|---|---|---|---|
| ln(стоимость экспорта) | −1.555 | 1.042 | 0.136 (n.s.) |
| ln(физический объём) | −0.332 | 0.882 | 0.706 (n.s.) |
| ln(удельная стоимость) | −0.122 | 0.264 | 0.645 (n.s.) |
| Доля в экспорте | −0.004 | 0.008 | 0.604 (n.s.) |

**Вывод:** Коэффициент при  статистически незначим по всем
зависимым переменным. HS 280530 (металлы РЗМ, апстрим) не демонстрирует значимой структурной
перестройки после реформы — в отличие от среднего передела (HS 284690).

### Тест предтрендов

| Тест | Статистика | p |
|---|---|---|
| Совместный F-тест (коэф. 2010–2013 = 0) | F = 0.155 | 0.959 |
| Непрерывный дифференциальный тренд 2010–2014 | β = 0.049 | 0.916 |

Предтренды параллельны в базовом периоде — предположение DID выполняется для данного кода,
хотя основной эффект незначим.
